src/modules/job-info/job_state.c

```diff
@@ -0,0 +1,479 @@
+        goto error;
+    zhashx_set_destructor (jsctx->index, job_destroy_wrapper);
+    zhashx_set_key_hasher (jsctx->index, job_hasher);
+    zhashx_set_key_comparator (jsctx->index, job_hash_key_cmp);
+    zhashx_set_key_duplicator (jsctx->index, NULL);
+    zhashx_set_key_destructor (jsctx->index, NULL);
+
+    if (!(jsctx->pending = zlistx_new ()))
+        goto error;
+    zlistx_set_comparator (jsctx->pending, job_list_cmp);
+
+    if (!(jsctx->running = zlistx_new ()))
+        goto error;
+
+    if (!(jsctx->inactive = zlistx_new ()))
+        goto error;
+
+    if (!(jsctx->processing = zlistx_new ()))
+        goto error;
+
+    if (!(jsctx->futures = zlist_new ()))
+        goto error;
+
+    if (flux_event_subscribe (h, "job-state") < 0) {
+        flux_log_error (h, "flux_event_subscribe");
+        goto error;
+    }
+
+    return jsctx;
+
+error:
+    saved_errno = errno;
+    job_state_destroy (jsctx);
+    errno = saved_errno;
+    return NULL;
+}
+
+void job_state_destroy (void *data)
+{
+    struct job_state_ctx *jsctx = data;
+    if (jsctx) {
+        /* Don't destroy processing until futures are complete */
+        if (jsctx->futures) {
+            flux_future_t *f;
+            f = zlist_pop (jsctx->futures);
+            while (f) {
+                if (flux_future_get (f, NULL) < 0)
+                    flux_log_error (jsctx->h, "%s: flux_future_get",
+                                    __FUNCTION__);
+                zlist_remove (jsctx->futures, f);
+                zlist_pop (jsctx->futures);
+            }
+            zlist_destroy (&jsctx->futures);
+        }
+        /* Destroy index last, as it is the one that will actually
+         * destroy the job objects */
+        if (jsctx->processing)
+            zlistx_destroy (&jsctx->processing);
+        if (jsctx->inactive)
+            zlistx_destroy (&jsctx->inactive);
+        if (jsctx->running)
+            zlistx_destroy (&jsctx->running);
+        if (jsctx->pending)
+            zlistx_destroy (&jsctx->pending);
+        if (jsctx->index)
+            zhashx_destroy (&jsctx->index);
+        free (jsctx);
+    }
+}
+
+/* zlistx_insert() and zlistx_reorder() take a 'low_value' parameter
+ * which indicates which end of the list to search from.
+ * false=search begins at tail (lowest priority, youngest)
+ * true=search begins at head (highest priority, oldest)
+ * Attempt to minimize search distance based on job priority.
+ */
+static bool search_direction (struct job *job)
+{
+    if (job->priority > FLUX_JOB_PRIORITY_DEFAULT)
+        return true;
+    else
+        return false;
+}
+
+/* remove job from one list and move it to another based on the
+ * newstate */
+static void job_change_list (struct job_state_ctx *jsctx,
+                             struct job *job,
+                             zlistx_t *oldlist,
+                             flux_job_state_t newstate)
+{
+    if (zlistx_delete (oldlist, job->list_handle) < 0)
```

Comment:
even though there's no destructor set on this list, `zlistx_detach()` might be a better choice here for clarity.